engine/runtime/src/main/java/org/enso/compiler/Cache.java

```diff
@@ -276,28 +297,19 @@ private Optional<T> loadCacheFrom(
   }
 
   /**
-   * Flag indicating if the de-serialization process should compute the hash of the sources from
-   * which the cache was created and compare it with the stored metadata entry.
-   */
-  protected abstract boolean needsSourceDigestVerification();
-
-  /**
-   * Flag indicating if the de-serialization process should compute the hash of the stored cache and
-   * compare it with the stored metadata entry.
-   */
-  protected abstract boolean needsDataDigestVerification();
-
-  /**
-   * Validates the deserialized data by returning the expected cached entry, or [[null]].
+   * Deserializes and validates data by returning the expected cached entry, or {@code null}.
    *
-   * @param obj deserialized object
+   * @param context the context
+   * @param data data to deserialize object from
    * @param meta metadata corresponding to the `obj`
    * @param logger Truffle's logger
-   * @return an `obj` transformed to a cached entry
+   * @return {@code data} transformed to a cached entry or {@code null}
    * @throws CacheException exception thrown on unexpected deserialized data
+   * @throws IOException when I/O goes wrong
+   * @throws ClassNotFoundException on problems with deserializaiton of Java classes
    */
-  protected abstract T validateReadObject(Object obj, M meta, TruffleLogger logger)
-      throws CacheException;
+  protected abstract T deserialize(EnsoContext context, byte[] data, M meta, TruffleLogger logger)
```

Comment:
With the `serialize` and `deserialize` methods @4e6 would be allowed to use JSON as the cache storage format for #5068